src/membership/view.rs

```diff
@@ -1,15 +1,295 @@
+use crate::{
+    common::{ConfigId, Endpoint, NodeId},
+    error::{Error, Result},
+};
+use std::collections::HashSet;
+
+type Ring = crate::membership::ring::Ring<Endpoint>;
+
 #[derive(Debug, Clone)]
 pub struct View {
-    configuration: Configuration,
+    k: i32,
+    rings: Vec<Ring>,
+    seen: HashSet<NodeId>,
+    current_config: ConfigId,
+    should_update_configuration_id: bool,
+    current_configuration: Configuration,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Default)]
 struct Configuration {
-    _p: (),
+    node_ids: Vec<NodeId>,
+    endpoints: Vec<Endpoint>,
 }
 
 impl View {
-    pub fn new() -> Self {
+    pub fn new(k: i32) -> Self {
+        assert!(k > 0);
+
+        let mut rings = Vec::with_capacity(k as usize);
+
+        for i in 0..k {
+            rings.push(Ring::new(i as u64));
+        }
+
+        Self {
+            k,
+            rings,
+            seen: HashSet::new(),
+            current_config: -1,
+            current_configuration: Configuration::default(),
+            should_update_configuration_id: true,
+        }
+    }
+
+    pub fn ring_add(&mut self, node: Endpoint, node_id: NodeId) -> Result<()> {
+        if self.is_node_present(&node_id) {
+            return Err(Error::new_uuid_already_seen());
+        }
+
+        if self.rings[0].contains(node.clone()) {
```

Comment:
So I'm going to make a separate comment about this, but I think it's important to document why checking only the first ring is fine. Reading the paper would tell you that all rings have the same nodes, so checking one is enough. But you need to remember this context.

Reply:
The reason this is safe is because when you create the view `K > 1` meaning that there is atleast one ring.